Simplify `SnippetViewSet` registration code

laymonage opened the pull request:
<!-- Thanks for contributing to Wagtail! 🎉  Please add a description below, explaining the purpose of this pull request - including the issue number of the issue you're fixing (if applicable). -->

Part of #10740. Built on top of #10767 to make rebasing easier later.

This simplifies the registration of `SnippetViewSet`s by:
- Making the `ViewSetRegistry` accept `ViewSetGroup` instead of only `ViewSet` instances
- Moving all side-effects of snippets registration to `SnippetViewSet.on_register()`
- Simplifying the underlying implementation of `register_snippet()`

With this change, if we want to have a similar registration mechanism for `{Foo}ViewSet`s, it's literally just:

```python
from wagtail.admin.viewsets import viewsets


def register_viewset(viewset_or_group):
    if callable(viewset_or_group):
        viewset_or_group = viewset_or_group()
    viewsets.register(viewset_or_group)
```

If we want `register_viewset` to strictly only allow classes instead of instances, we could get rid of the `if` check and make it a one-line function.

To use it, you would do:

```python
register_viewset(MyViewSet)
# or
register_viewset(MyViewSet(...))
```

Which is basically a more unified replacement for the existing `register_admin_viewset` hook (which probably should be deprecated at this point?)

```python
@hooks.register("register_admin_viewset")
def register_myviewset():
    return MyViewSet()
```

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

laymonage commented on wagtail/snippets/models.py:
```diff
@@ -83,51 +83,33 @@ def register_snippet(registerable, viewset=None):
 def _register_snippet_immediately(registerable, viewset=None):
```
This could be further simplified if we stopped supporting the following registration style:

```py
register_snippet(MyModel, viewset=MyModelViewSet)
```

And instead just support the following styles:

```py
# Old style:
register_snippet(MyModel)
# equivalent to
@register_snippet
class MyModel(models.Model):
    ....

# With custom ViewSet
register_snippet(MyModelViewSet)

# Optionally, we can also support instances, not just classes
# (this is already done in this PR)
register_snippet(SomeGenericViewSet(model=MyModel, ...))
```

The `viewset=` parameter was only documented for some time when custom `SnippetViewSet` support was added but before the custom menu registration was added, so 4.1.x and 4.2.x releases. As of 5.0.x, the docs no longer mentions `viewset=` in favour of passing the viewset to `register_snippet` as the first parameter and setting the model on the viewset.

laymonage reviewed (commented):
This is ready for review once #10767 is merged, or both can be reviewed at once in this PR.

gasman commented on wagtail/admin/viewsets/model.py:
```diff
@@ -195,6 +199,7 @@ def get_urlpatterns(self):
     def on_register(self):
         super().on_register()
         self.register_admin_url_finder()
+        self.register_reference_index()
```
Worth noting that this will enable reference tracking for the handful of models that have a ModelViewSet within Wagtail as standard - that would be Sites, Collections, Locales, Users and Groups I think. I don't think there are any negative consequences to that, though - after all, we were indexing those (along with all other models project-wide) in Wagtail 4.1 - 4.2.

laymonage replied:
Good catch, thanks! I think it's worth adding an `add_to_reference_index` flag to the `ModelViewSet`. I also think we should turn indexing off for those models at least for now, in case that would slow down instances that have many foreign keys to those models. The models specifically are Site, Locale, and Group, as the Collection and User views haven't been migrated to use `ModelViewSet`.

I've done so and added upgrade considerations & release notes.

gasman reviewed (approved):
Looks good to me! Just one observation about moving ReferenceIndex registration to ModelViewSet, which I don't think is a blocker.

laymonage commented:
Hmmm, not sure what's going on with the CI here... pretty sure my changes since the last commit were harmless. The tests still pass on my local machine. The error seems really weird though, I've no idea why it complains about having duplicate `wagtailcore`-labelled apps.

laymonage commented:
Tried running with a fresh venv and installing the dependencies with `python -m pip install -e ".[testing]" --no-cache-dir --upgrade` locally, no luck in reproducing the CI failures. Tried busting the cache on GitHub Actions too, and that didn't work.

CircleCI seems to be running the tests just fine, like on my machine. Really don't know what the issue is here... 😕

laymonage commented:
OK something is definitely up with GitHub Actions, I added a commit literally reverts all the commits since the previous successful run, and it still fails. I'm going to revert the revert and merge it anyway then...
